ENH: Add support for building against VTK version 9.5.0

jamesobutler opened the pull request:
This updates Slicer to add building against VTK 9.5.0 which is a culmination of 6 months of work. VTK is now following a 6 month cadence of releases as described in the 9.5.0 release process [announcement](https://discourse.vtk.org/t/vtk-9-5-0-release/15018). See full diff https://github.com/Kitware/VTK/compare/v9.4.2...v9.5.0.rc1.

Read https://gitlab.kitware.com/vtk/vtk/-/tree/master/Documentation/release/9.5?ref_type=heads for specific details on 9.5.0 changes.
- VTK is now enforcing minimum C++17 and therefore now supports the following compilers:
  - GCC 8.0 or newer
  - Clang 5.0 or newer
  - Apple Clang 10.0 or newer
  - Microsoft Visual Studio 2017 or newer
  - Intel 19.0 or newer

------------------------------------------
This PR is currently building VTK based off of 9.5.0.rc1 to participate in the release candidate period as announced [here](https://discourse.vtk.org/t/vtk-9-5-0-rc1/15630). A Slicer forked branch based off VTK 9.5.0.rc1 is being used in this PR as can be seen at [`slicer-v9.5.0.rc1-2025-05-14-fe7cd0efa`](https://github.com/Slicer/VTK/tree/slicer-v9.5.0.rc1-2025-05-14-fe7cd0efa).

- [x] https://github.com/Slicer/vtkAddon/pull/56 integration will be necessary
- [x] Update the vtkAddon repository back to the upstream and update the git hash following the integration of the above step. See https://github.com/Slicer/Slicer/pull/8429

Additionally utilizing latest VTK 9.5.0 release candidate code may be beneficial towards coming to a resolution regarding issues building debug builds as described in https://discourse.vtk.org/t/vtk-9-4-debug-build-fails-with-msvc-due-to-image-size-exceeds-maximum-allowable-size/15520.

FYI @sankhesh, that this PR will indicate how well the VTK 9.5.0 RC is working in the context of 3D Slicer. 

------------------------------------------
3D Slicer is currently building and packaging successfully with this branch on the Windows platform.

- [x] There are 4 newly failing automated tests with this branch that all appear to be having the same issue where the app is exiting abnormally upon calls such as `int retval = vtkRegressionTestImage(renderWindow);`.
- `vtkMRMLCameraDisplayableManagerTest1`
- `vtkMRMLModelClipDisplayableManagerTest`
- `vtkMRMLModelDisplayableManagerTest`
- `vtkMRMLThreeDReformatDisplayableManagerTest1`

lassoan commented:
Thanks for working on this!

We'll also need to include this VTK fix that allows debug builds on Windows: https://gitlab.kitware.com/vtk/vtk/-/merge_requests/12133

jamesobutler commented:
Great! I didn't see that there was already a debug build fix. I'll definitely add to the [`slicer-v9.5.0.rc1-2025-05-14-fe7cd0efa`](https://github.com/Slicer/VTK/tree/slicer-v9.5.0.rc1-2025-05-14-fe7cd0efa) branch.

jamesobutler commented:
@lassoan I've updated the Slicer VTK fork branch and updated the git SHA in this branch to use it (https://github.com/Slicer/VTK/commit/993649f58a7ced5822a6556321e747d3b128211a). So you should be able to try debug builds with this PR.

sankhesh commented on Libs/MRML/DisplayableManager/Testing/Cxx/vtkMRMLCameraDisplayableManagerTest1.cxx:
```diff
@@ -708,7 +708,9 @@ int vtkMRMLCameraDisplayableManagerTest1(int argc, char* argv[])
-  int retval = vtkRegressionTestImageThreshold(rw.GetPointer(), 85.0);
+  // int retval = vtkRegressionTestImageThreshold(rw.GetPointer(), 85.0);
+  /// failure above line
+  int retval = vtkRegressionTester::DO_INTERACTOR;
```
vtkRegressionTestImage has changed to use vtkImageSSIM metrics for image difference based testing. Maybe that is why there are failures :shrug:  You could set up to test using the legacy method or update the thresholds/baselines. See https://gitlab.kitware.com/vtk/vtk/-/blob/a7da845df6bf1c63945218512b5c5926b03fbf17/Testing/Rendering/vtkTesting.cxx#L718-737

jamesobutler commented:
@lassoan Following up on the ability to build a debug build - I was successfully able to build this branch in Debug mode on the Windows platform.

jamesobutler commented:
I've updated this branch pointing to latest VTK release code for the 9.5.0 release. It is currently in a period post rc2 with final release estimated for June 17th per https://discourse.vtk.org/t/vtk-9-5-0-release/15018/6.

lassoan commented:
Thanks a lot, I've tested the build on Windows and it worked well. It would be great if you could rebase on the latest main and check if using the legacy image compare method as Sankhesh suggested fixes the problem (or we could update the regression image, if the current behavior looks good).

jamesobutler commented:
I have picked off my debugging commit here associated with failing tests that use `vtkRegressionTestImageThreshold`. These tests include:
- vtkMRMLCameraDisplayableManagerTest1
- vtkMRMLModelDisplayableManagerTest1
- vtkMRMLThreeDReformatDisplayableManagerTest1
- vtkMRMLVolumeRenderingDisplayableManagerTest1

lassoan commented:
@jamesobutler this seems to work well now. If everything looks good to you, too, then please merge. Thanks a lot!

jamesobutler commented:
Thanks I have tested locally and confirmed those last few testing issued have been resolved.

lassoan commented:
Awesome, thank you!
